server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java

```diff
@@ -122,22 +122,27 @@ public boolean isReplicated() {
      * Performs a rollover on a {@code DataStream} instance and returns a new instance containing
      * the updated list of backing indices and incremented generation.
      *
-     * @param newWriteIndex the new write backing index. Must conform to the naming convention for
-     *                      backing indices on data streams. See {@link #getDefaultBackingIndexName}.
+     * @param writeIndexUuid UUID for the new write index
+     * @param  minNodeVersion minimum cluster node version
+     *
      * @return new {@code DataStream} instance with the rollover operation applied
      */
-    public DataStream rollover(Index newWriteIndex, Version minNodeVersion) {
-        assert newWriteIndex.getName().equals(getDefaultBackingIndexName(name, generation + 1, minNodeVersion));
+    public DataStream rollover(String writeIndexUuid, Version minNodeVersion) {
         if (replicated) {
             throw new IllegalArgumentException("data stream [" + name + "] cannot be rolled over, " +
                 "because it is a replicated data stream");
         }
 
         List<Index> backingIndices = new ArrayList<>(indices);
-        backingIndices.add(newWriteIndex);
+        final String newWriteIndexName = DataStream.getDefaultBackingIndexName(getName(), getGeneration() + 1, minNodeVersion);
+        backingIndices.add(new Index(newWriteIndexName, writeIndexUuid));
         return new DataStream(name, timeStampField, backingIndices, generation + 1, metadata, hidden, replicated);
     }
 
+    public DataStream rollover(Version minNodeVersion) {
+        return rollover("uuid", minNodeVersion);
```

Comment:
Maybe we can always use `UUIDs.randomBase64UUID()` as uuid for write index id and then we don't need to overload the rollover method? 

I think we then only need to adjust tests when asserting a data stream, but I think that is cleaner. We could make a custom hamcrest matcher to make comparing easier.

Reply:
I can make that change but then the UUID for the new write index will not be known in advance. `MetadataRolloverService::rolloverDataStream` will create the new write index and then call `DataStream::rollover` which will add the new index to the data stream by name but with a different random UUID. Is that acceptable?

 and the assertion on line 1391 in `Metadata::buildIndicesLookup` will fail unless I change it to look only for the index name. Do you think that's acceptable

Reply:
I think so. Then the changes in `ReactiveStorageDeciderService.java` will not need to overloaded version

Reply:
Thanks, @martijnvg. I've made those changes.